providers/structured/provider.go

```diff
@@ -55,68 +73,71 @@ func NewStructuredOutputLogProvider() *StructuredOutputLogProvider {
 }
 
 func LogProvider(nextProvider providers.LogProvider) *StructuredOutputLogProvider {
-	rawLogCalls := map[providers.RawLogCallType][]RawLogCallArgs{}
-	for _, callType := range providers.RawLogCallTypes() {
-		rawLogCalls[callType] = []RawLogCallArgs{}
-	}
-
 	return &StructuredOutputLogProvider{
-		LogProvider:      chaining.LogProvider(nextProvider),
-		rawLogCalls:      rawLogCalls,
-		recordCalls:      []RecordCallArgs{},
-		recordEventCalls: []RecordEventCallArgs{},
+		LogProvider: chaining.LogProvider(nextProvider),
+		logCalls:    []*LogCallArgs{},
+		recordCalls: []*RecordCallArgs{},
 	}
 }
 
-func (p *StructuredOutputLogProvider) saveRawCallArgs(
-	callType providers.RawLogCallType,
+func (p *StructuredOutputLogProvider) saveLogCallArgs(
+	level providers.LogLevel,
 	ctx context.Context,
 	report bool,
 	args ...interface{},
 ) {
-	callArgs := RawLogCallArgs{
+	callArgs := LogCallArgs{
 		ContextFields: log.FieldsFromContext(ctx),
 		Report:        report,
 		Args:          args,
+		Level:         level,
 	}
-	p.rawLogCalls[callType] = append(p.rawLogCalls[callType], callArgs)
+	p.logMutex.Lock()
+	defer p.logMutex.Unlock()
+	p.logCalls = append(p.logCalls, &callArgs)
 }
 
 func (p *StructuredOutputLogProvider) Error(ctx context.Context, report bool, args ...interface{}) {
-	p.saveRawCallArgs(providers.Error, ctx, report, args...)
+	p.saveLogCallArgs(providers.Error, ctx, report, args...)
 	p.LogProvider.Error(ctx, report, args...)
 }
 
 func (p *StructuredOutputLogProvider) Warn(ctx context.Context, report bool, args ...interface{}) {
-	p.saveRawCallArgs(providers.Warn, ctx, report, args...)
+	p.saveLogCallArgs(providers.Warn, ctx, report, args...)
 	p.LogProvider.Warn(ctx, report, args...)
 }
 
 func (p *StructuredOutputLogProvider) Info(ctx context.Context, report bool, args ...interface{}) {
-	p.saveRawCallArgs(providers.Info, ctx, report, args...)
+	p.saveLogCallArgs(providers.Info, ctx, report, args...)
 	p.LogProvider.Info(ctx, report, args...)
 }
 
 func (p *StructuredOutputLogProvider) Debug(ctx context.Context, report bool, args ...interface{}) {
-	p.saveRawCallArgs(providers.Debug, ctx, report, args...)
+	p.saveLogCallArgs(providers.Debug, ctx, report, args...)
 	p.LogProvider.Debug(ctx, report, args...)
 }
 
 func (p *StructuredOutputLogProvider) Record(ctx context.Context, metrics map[string]interface{}) {
+	p.LogProvider.Record(ctx, metrics)
```

Comment:
We can call the next provider first, i.e. the order doesn't matter then?

Reply:
Can't see a reason the order matters here, so I put this first to reduce time holding the lock.